plugins/resource_monitor_plugin/resource_monitor_plugin.cpp

```diff
@@ -13,45 +13,41 @@
    The plugin uses a thread to periodically check space usage of file
    systems of directories being monitored. If space used
    is over a predefined threshold, a graceful shutdown is initiated.
-
-   The directories to be monitored are configured;
-   they can be any combination of data, data/blocks, data/state and
-   any directories like those for trace and state history.
-   Space threshold can also be configured.
-
-   Alternative solution to blocks directory space exhaustion
-   was considered: starting from block_log::append, 
-   make sure all exceptions was handled properly. But it was deemed to be
-   too tedious and error prone to review all possible excpetions, and
-   might likely break concensus. The plugin approach is more generic,
-   monitoring any file systems and without modifying existing code.
-
 **/
 
 #include <eosio/resource_monitor_plugin/resource_monitor_plugin.hpp>
+#include <eosio/resource_monitor_plugin/file_space_handler.hpp>
+#include <eosio/resource_monitor_plugin/system_file_space_provider.hpp>
+
 #include <eosio/chain/exceptions.hpp>
 
 #include <fc/exception/exception.hpp>
 #include <fc/log/logger_config.hpp> // set_os_thread_name
 
-#include <boost/filesystem.hpp> // C++17's filesystem not in our toolchain, use Boost's
+#include <boost/filesystem.hpp>
 
 #include <thread>
 
 #include <sys/stat.h>
 
+using namespace eosio::resource_monitor;
+
 namespace bfs = boost::filesystem;
 
 namespace eosio {
    static appbase::abstract_plugin& _resource_monitor_plugin = app().register_plugin<resource_monitor_plugin>();
 
+   static vector<bfs::path> directories_monitored;
```

Comment:
This should be a member of `resource_monitor_plugin_impl` and not a static.
`monitor_directory()` can find the plugin and if available register the directory.

Reply:
Thank you so much Kevin for detailed comments! I will incorporate them.

Will use find_plugin to get the API.